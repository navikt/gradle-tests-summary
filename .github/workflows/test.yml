name: Test action

on: push

jobs:
  my-job:
    name: my-job
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: oven-sh/setup-bun@v2
      - run: bun i
      - run: bun test
      - run: bun run build
