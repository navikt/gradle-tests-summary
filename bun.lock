{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "gradle-tests-summary",
      "dependencies": {
        "@actions/core": "^1.11.1",
        "@actions/github": "^6.0.0",
      },
      "devDependencies": {
        "@karl-run/prettier-config": "^1.0.1",
        "@types/bun": "latest",
        "prettier": "^3.5.3",
      },
      "peerDependencies": {
        "typescript": "^5",
      },
    },
  },
  "packages": {
    "@actions/core": ["@actions/core@1.11.1", "", { "dependencies": { "@actions/exec": "^1.1.1", "@actions/http-client": "^2.0.1" } }, "sha512-hXJCSrkwfA46Vd9Z3q4cpEpHB1rL5NG04+/rbqW9d3+CSvtB1tYe8UTpAlixa1vj0m/ULglfEK2UKxMGxCxv5A=="],

    "@actions/exec": ["@actions/exec@1.1.1", "", { "dependencies": { "@actions/io": "^1.0.1" } }, "sha512-+sCcHHbVdk93a0XT19ECtO/gIXoxvdsgQLzb2fE2/5sIZmWQuluYyjPQtrtTHdU1YzTZ7bAPN4sITq2xi1679w=="],

    "@actions/github": ["@actions/github@6.0.0", "", { "dependencies": { "@actions/http-client": "^2.2.0", "@octokit/core": "^5.0.1", "@octokit/plugin-paginate-rest": "^9.0.0", "@octokit/plugin-rest-endpoint-methods": "^10.0.0" } }, "sha512-alScpSVnYmjNEXboZjarjukQEzgCRmjMv6Xj47fsdnqGS73bjJNDpiiXmp8jr0UZLdUB6d9jW63IcmddUP+l0g=="],

    "@actions/http-client": ["@actions/http-client@2.2.3", "", { "dependencies": { "tunnel": "^0.0.6", "undici": "^5.25.4" } }, "sha512-mx8hyJi/hjFvbPokCg4uRd4ZX78t+YyRPtnKWwIl+RzNaVuFpQHfmlGVfsKEJN8LwTCvL+DfVgAM04XaHkm6bA=="],

    "@actions/io": ["@actions/io@1.1.3", "", {}, "sha512-wi9JjgKLYS7U/z8PPbco+PvTb/nRWjeoFlJ1Qer83k/3C5PHQi28hiVdeE2kHXmIL99mQFawx8qt/JPjZilJ8Q=="],

    "@fastify/busboy": ["@fastify/busboy@2.1.1", "", {}, "sha512-vBZP4NlzfOlerQTnba4aqZoMhE/a9HY7HRqoOPaETQcSQuWEIyZMHGfVu6w9wGtGK5fED5qRs2DteVCjOH60sA=="],

    "@karl-run/prettier-config": ["@karl-run/prettier-config@1.0.1", "", { "peerDependencies": { "prettier": "*" } }, "sha512-KsQktHGFGM7gb+F2ba4U4KioWfSRRx8sp7+GoDVFSl/u7e5JKefyrPnKZhEi6gz7YLKGbYzqKJlSGwpz+CNyZA=="],

    "@octokit/auth-token": ["@octokit/auth-token@4.0.0", "", {}, "sha512-tY/msAuJo6ARbK6SPIxZrPBms3xPbfwBrulZe0Wtr/DIY9lje2HeV1uoebShn6mx7SjCHif6EjMvoREj+gZ+SA=="],

    "@octokit/core": ["@octokit/core@5.2.0", "", { "dependencies": { "@octokit/auth-token": "^4.0.0", "@octokit/graphql": "^7.1.0", "@octokit/request": "^8.3.1", "@octokit/request-error": "^5.1.0", "@octokit/types": "^13.0.0", "before-after-hook": "^2.2.0", "universal-user-agent": "^6.0.0" } }, "sha512-1LFfa/qnMQvEOAdzlQymH0ulepxbxnCYAKJZfMci/5XJyIHWgEYnDmgnKakbTh7CH2tFQ5O60oYDvns4i9RAIg=="],

    "@octokit/endpoint": ["@octokit/endpoint@9.0.6", "", { "dependencies": { "@octokit/types": "^13.1.0", "universal-user-agent": "^6.0.0" } }, "sha512-H1fNTMA57HbkFESSt3Y9+FBICv+0jFceJFPWDePYlR/iMGrwM5ph+Dd4XRQs+8X+PUFURLQgX9ChPfhJ/1uNQw=="],

    "@octokit/graphql": ["@octokit/graphql@7.1.1", "", { "dependencies": { "@octokit/request": "^8.4.1", "@octokit/types": "^13.0.0", "universal-user-agent": "^6.0.0" } }, "sha512-3mkDltSfcDUoa176nlGoA32RGjeWjl3K7F/BwHwRMJUW/IteSa4bnSV8p2ThNkcIcZU2umkZWxwETSSCJf2Q7g=="],

    "@octokit/openapi-types": ["@octokit/openapi-types@23.0.1", "", {}, "sha512-izFjMJ1sir0jn0ldEKhZ7xegCTj/ObmEDlEfpFrx4k/JyZSMRHbO3/rBwgE7f3m2DHt+RrNGIVw4wSmwnm3t/g=="],

    "@octokit/plugin-paginate-rest": ["@octokit/plugin-paginate-rest@9.2.2", "", { "dependencies": { "@octokit/types": "^12.6.0" }, "peerDependencies": { "@octokit/core": "5" } }, "sha512-u3KYkGF7GcZnSD/3UP0S7K5XUFT2FkOQdcfXZGZQPGv3lm4F2Xbf71lvjldr8c1H3nNbF+33cLEkWYbokGWqiQ=="],

    "@octokit/plugin-rest-endpoint-methods": ["@octokit/plugin-rest-endpoint-methods@10.4.1", "", { "dependencies": { "@octokit/types": "^12.6.0" }, "peerDependencies": { "@octokit/core": "5" } }, "sha512-xV1b+ceKV9KytQe3zCVqjg+8GTGfDYwaT1ATU5isiUyVtlVAO3HNdzpS4sr4GBx4hxQ46s7ITtZrAsxG22+rVg=="],

    "@octokit/request": ["@octokit/request@8.4.1", "", { "dependencies": { "@octokit/endpoint": "^9.0.6", "@octokit/request-error": "^5.1.1", "@octokit/types": "^13.1.0", "universal-user-agent": "^6.0.0" } }, "sha512-qnB2+SY3hkCmBxZsR/MPCybNmbJe4KAlfWErXq+rBKkQJlbjdJeS85VI9r8UqeLYLvnAenU8Q1okM/0MBsAGXw=="],

    "@octokit/request-error": ["@octokit/request-error@5.1.1", "", { "dependencies": { "@octokit/types": "^13.1.0", "deprecation": "^2.0.0", "once": "^1.4.0" } }, "sha512-v9iyEQJH6ZntoENr9/yXxjuezh4My67CBSu9r6Ve/05Iu5gNgnisNWOsoJHTP6k0Rr0+HQIpnH+kyammu90q/g=="],

    "@octokit/types": ["@octokit/types@13.8.0", "", { "dependencies": { "@octokit/openapi-types": "^23.0.1" } }, "sha512-x7DjTIbEpEWXK99DMd01QfWy0hd5h4EN+Q7shkdKds3otGQP+oWE/y0A76i1OvH9fygo4ddvNf7ZvF0t78P98A=="],

    "@types/bun": ["@types/bun@1.2.5", "", { "dependencies": { "bun-types": "1.2.5" } }, "sha512-w2OZTzrZTVtbnJew1pdFmgV99H0/L+Pvw+z1P67HaR18MHOzYnTYOi6qzErhK8HyT+DB782ADVPPE92Xu2/Opg=="],

    "@types/node": ["@types/node@22.13.10", "", { "dependencies": { "undici-types": "~6.20.0" } }, "sha512-I6LPUvlRH+O6VRUqYOcMudhaIdUVWfsjnZavnsraHvpBwaEyMN29ry+0UVJhImYL16xsscu0aske3yA+uPOWfw=="],

    "@types/ws": ["@types/ws@8.5.14", "", { "dependencies": { "@types/node": "*" } }, "sha512-bd/YFLW+URhBzMXurx7lWByOu+xzU9+kb3RboOteXYDfW+tr+JZa99OyNmPINEGB/ahzKrEuc8rcv4gnpJmxTw=="],

    "before-after-hook": ["before-after-hook@2.2.3", "", {}, "sha512-NzUnlZexiaH/46WDhANlyR2bXRopNg4F/zuSA3OpZnllCUgRaOF2znDioDWrmbNVsuZk6l9pMquQB38cfBZwkQ=="],

    "bun-types": ["bun-types@1.2.5", "", { "dependencies": { "@types/node": "*", "@types/ws": "~8.5.10" } }, "sha512-3oO6LVGGRRKI4kHINx5PIdIgnLRb7l/SprhzqXapmoYkFl5m4j6EvALvbDVuuBFaamB46Ap6HCUxIXNLCGy+tg=="],

    "deprecation": ["deprecation@2.3.1", "", {}, "sha512-xmHIy4F3scKVwMsQ4WnVaS8bHOx0DmVwRywosKhaILI0ywMDWPtBSku2HNxRvF7jtwDRsoEwYQSfbxj8b7RlJQ=="],

    "once": ["once@1.4.0", "", { "dependencies": { "wrappy": "1" } }, "sha512-lNaJgI+2Q5URQBkccEKHTQOPaXdUxnZZElQTZY0MFUAuaEqe1E+Nyvgdz/aIyNi6Z9MzO5dv1H8n58/GELp3+w=="],

    "prettier": ["prettier@3.5.3", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-QQtaxnoDJeAkDvDKWCLiwIXkTgRhwYDEQCghU9Z6q03iyek/rxRh/2lC3HB7P8sWT2xC/y5JDctPLBIGzHKbhw=="],

    "tunnel": ["tunnel@0.0.6", "", {}, "sha512-1h/Lnq9yajKY2PEbBadPXj3VxsDDu844OnaAo52UVmIzIvwwtBPIuNvkjuzBlTWpfJyUbG3ez0KSBibQkj4ojg=="],

    "typescript": ["typescript@5.8.2", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-aJn6wq13/afZp/jT9QZmwEjDqqvSGp1VT5GVg+f/t6/oVyrgXM6BY1h9BRh/O5p3PlUPAe+WuiEZOmb/49RqoQ=="],

    "undici": ["undici@5.28.5", "", { "dependencies": { "@fastify/busboy": "^2.0.0" } }, "sha512-zICwjrDrcrUE0pyyJc1I2QzBkLM8FINsgOrt6WjA+BgajVq9Nxu2PbFFXUrAggLfDXlZGZBVZYw7WNV5KiBiBA=="],

    "undici-types": ["undici-types@6.20.0", "", {}, "sha512-Ny6QZ2Nju20vw1SRHe3d9jVu6gJ+4e3+MMpqu7pqE5HT6WsTSlce++GQmK5UXS8mzV8DSYHrQH+Xrf2jVcuKNg=="],

    "universal-user-agent": ["universal-user-agent@6.0.1", "", {}, "sha512-yCzhz6FN2wU1NiiQRogkTQszlQSlpWaw8SvVegAc+bDxbzHgh1vX8uIe8OYyMH6DwH+sdTJsgMl36+mSMdRJIQ=="],

    "wrappy": ["wrappy@1.0.2", "", {}, "sha512-l4Sp/DRseor9wL6EvV2+TuQn63dMkPjZ/sp9XkghTEbV9KlPS1xUsZ3u7/IQO4wxtcFB4bgpQPRcR3QCvezPcQ=="],

    "@octokit/plugin-paginate-rest/@octokit/types": ["@octokit/types@12.6.0", "", { "dependencies": { "@octokit/openapi-types": "^20.0.0" } }, "sha512-1rhSOfRa6H9w4YwK0yrf5faDaDTb+yLyBUKOCV4xtCDB5VmIPqd/v9yr9o6SAzOAlRxMiRiCic6JVM1/kunVkw=="],

    "@octokit/plugin-rest-endpoint-methods/@octokit/types": ["@octokit/types@12.6.0", "", { "dependencies": { "@octokit/openapi-types": "^20.0.0" } }, "sha512-1rhSOfRa6H9w4YwK0yrf5faDaDTb+yLyBUKOCV4xtCDB5VmIPqd/v9yr9o6SAzOAlRxMiRiCic6JVM1/kunVkw=="],

    "@octokit/plugin-paginate-rest/@octokit/types/@octokit/openapi-types": ["@octokit/openapi-types@20.0.0", "", {}, "sha512-EtqRBEjp1dL/15V7WiX5LJMIxxkdiGJnabzYx5Apx4FkQIFgAfKumXeYAqqJCj1s+BMX4cPFIFC4OLCR6stlnA=="],

    "@octokit/plugin-rest-endpoint-methods/@octokit/types/@octokit/openapi-types": ["@octokit/openapi-types@20.0.0", "", {}, "sha512-EtqRBEjp1dL/15V7WiX5LJMIxxkdiGJnabzYx5Apx4FkQIFgAfKumXeYAqqJCj1s+BMX4cPFIFC4OLCR6stlnA=="],
  }
}
