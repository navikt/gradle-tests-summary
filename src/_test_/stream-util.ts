import { Writable } from 'stream'

export function captureWritable(): { stream: Writable; getOutput: () => string } {
  let output = ''

  const stream = new Writable({
    write(chunk, _, callback) {
      output += chunk.toString()
      callback()
    },
  })

  return { stream, getOutput: () => output }
}
