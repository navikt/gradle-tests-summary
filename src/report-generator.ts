import type { Writable } from 'node:stream'
import * as fs from 'node:fs'
import * as path from 'node:path'

import * as glob from 'glob'
import { DOMParser } from 'xmldom'
import * as xpath from 'xpath'

type CountKeys = 'tests' | 'passed' | 'skipped' | 'failures' | 'errors'

const COUNT_XPATHS: Record<CountKeys, string> = {
  tests: '//testcase',
  passed: '//testcase[not(*)]',
  skipped: '//testcase[skipped]',
  failures: '//testcase[failure]',
  errors: '//testcase[error]',
}

export class ReportGenerator {
  constructor(private pathPrefix: string) {}

  generateReport(out: Writable) {
    const testFiles = glob.sync(`${this.pathPrefix}**/{TEST,test}-*.xml`)
    const grouped = this.groupBySubproject(testFiles)

    // Determine the widest subproject name
    const subprojects = Object.keys(grouped)
    const subprojectWidth = Math.max('Subproject'.length, ...subprojects.map((p) => p.length))
    const isSingleSubproject = subprojects.length === 1

    this.printHeaders(out, subprojectWidth, isSingleSubproject)

    for (const [group, testResults] of Object.entries(grouped)) {
      this.process(out, group, testResults, subprojectWidth, isSingleSubproject)
    }
  }

  private groupBySubproject(files: string[]): Record<string, string[]> {
    return files.reduce(
      (acc, file) => {
        const relativePath = file.slice(this.pathPrefix.length)
        const group = relativePath.split(path.sep)[0]
        acc[group] = acc[group] || []
        acc[group].push(file)
        return acc
      },
      {} as Record<string, string[]>,
    )
  }

  private process(
    out: Writable,
    group: string,
    testResults: string[],
    subprojectWidth: number,
    isSingleSubproject: boolean,
  ) {
    const docs = testResults.map((file) => {
      const content = fs.readFileSync(file, 'utf-8')
      return new DOMParser({
        errorHandler: (level, msg) => {
          if (level === 'fatalError' || process.env.VERBOSE === 'true') {
            console.error(msg)
          }
        },
      }).parseFromString(content, 'text/xml')
    })

    this.printGroup(out, group, this.counts(docs), subprojectWidth, isSingleSubproject)
  }

  private counts(docs: Document[]): Record<CountKeys, number> {
    return Object.fromEntries(
      Object.entries(COUNT_XPATHS).map(([key, xpathQuery]) => [key, this.count(docs, xpathQuery)]),
    ) as Record<CountKeys, number>
  }

  private count(docs: Document[], xpathQuery: string): number {
    return docs.reduce((sum, doc) => sum + (xpath.select('count(' + xpathQuery + ')', doc) as number), 0)
  }

  private printHeaders(out: Writable, subprojectWidth: number, isSingleSubproject: boolean) {
    const subprojectCol = 'Subproject'.padEnd(subprojectWidth)
    out.write("# Test Results\n\n")
    if (isSingleSubproject) {
      out.write(`| Status | Tests | Passed | Skipped | Failures | Errors |\n`)
      out.write(`|:------:|:-----:|:------:|:-------:|:--------:|:------:|\n`)
    } else {
      out.write(`| ${subprojectCol} | Status | Tests | Passed | Skipped | Failures | Errors |\n`)
      out.write(`|-${'-'.repeat(subprojectWidth)}-|:------:|:-----:|:------:|:-------:|:--------:|:------:|\n`)
    }
  }

  private printGroup(
    out: Writable,
    group: string,
    counts: Record<CountKeys, number>,
    subprojectWidth: number,
    isSingleSubproject: boolean,
  ) {
    const status = counts.failures === 0 && counts.errors === 0 ? ' ✅ ' : ' ❌ '
    const paddedGroup = group.padEnd(subprojectWidth)

    const formatNumber = (num: number, headerWidth: number) => num.toString().padStart(headerWidth, ' ')

    const stats = [
      paddedGroup,
      status.padStart(5).padEnd(3),
      formatNumber(counts.tests, 5),
      formatNumber(counts.passed, 6),
      formatNumber(counts.skipped, 7),
      formatNumber(counts.failures, 8),
      formatNumber(counts.errors, 6),
    ]

    if (isSingleSubproject) {
      out.write(`| ${stats.slice(1).join(' | ')} |\n`)
    } else {
      out.write(`| ${stats.join(' | ')} |\n`)
    }
  }
}
