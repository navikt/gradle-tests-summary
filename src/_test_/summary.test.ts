import { test, expect } from 'bun:test'

import { ReportGenerator } from '../report-generator'
import { captureWritable } from './stream-util'

test('should handle single module', () => {
  const reportGenerator = new ReportGenerator('src/_test_/data/single-module/')

  const output = captureWritable()
  reportGenerator.generateReport(output.stream)

  expect(output.getOutput()).toMatchSnapshot()
})

test('should handle single module', () => {
  const reportGenerator = new ReportGenerator('src/_test_/data/multi-module/')

  const output = captureWritable()
  reportGenerator.generateReport(output.stream)

  expect(output.getOutput()).toMatchSnapshot()
})
