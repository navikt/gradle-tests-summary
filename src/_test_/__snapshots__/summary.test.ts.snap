// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle single module 1`] = `
"# Test Results

| Status | Tests | Passed | Skipped | Failures | Errors |
|:------:|:-----:|:------:|:-------:|:--------:|:------:|
|    ✅  |     6 |      6 |       0 |        0 |      0 |
"
`;

exports[`should handle single module 2`] = `
"# Test Results

| Subproject | Status | Tests | Passed | Skipped | Failures | Errors |
|------------|:------:|:-----:|:------:|:-------:|:--------:|:------:|
| module-b   |    ✅  |     6 |      6 |       0 |        0 |      0 |
| module-a   |    ✅  |     6 |      6 |       0 |        0 |      0 |
"
`;
